.travis.yml

```diff
@@ -10,29 +10,48 @@ services:
+  - OAUTHLIB_INSECURE=0
+  - USERS_ROLE=.users_role.json
+  - APP_LOGIN_DISABLED=false
+  - AZ_LOCATION=uksouth
 jobs:
   include:
   - stage: Build docker images
+    before_deploy:
+    - wget https://releases.hashicorp.com/terraform/$TF_VERSION/terraform_"$TF_VERSION"_linux_amd64.zip
+    - unzip terraform_"$TF_VERSION"_linux_amd64.zip
+    - sudo mv terraform /usr/local/bin/
+    - rm terraform_"$TF_VERSION"_linux_amd64.zip
+    - terraform init
+    - terraform apply -var "prefix=mmce-test2" -var "location=${AZ_LOCATION}" -var "flask_app=${FLASK_APP}" -var "docker_enable_ci=true" -var "docker_registry_server_password=${DOCKER_PASSWORD}" -var "docker_registry_server_url=${DOCKER_REGISTRY_URL}" -var "docker_registry_server_username=${DOCKER_USERNAME}" -var "github_client_id=${GITHUB_CLIENT_ID}" -var "github_secret=${GITHUB_SECRET}" -var "oauthlib_insecure=${OAUTHLIB_INSECURE}" -var "secret_key=${SECRET_KEY}" -var "users_role=${USERS_ROLE}" -var "default_database=${DEFAULT_DATABASE}" -var "login_disabled=${LOGIN_DISABLED}" -var "app_login_disabled=${APP_LOGIN_DISABLED}" -auto-approve
+    - AZURE_WEBHOOK=$(terraform output -raw webhook_url)
+    - MONGO_CONNECTION_STRING=`terraform output -raw mongo_connection_string | sed -r 's/\&+/\^\&+/g'`
     before_script:
     - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
     script:
-    - docker build --target production --tag $DOCKER_USERNAME/todo-app:latest .
-    - docker push $DOCKER_USERNAME/todo-app:latest
+      - docker build --target test --tag $DOCKER_USERNAME/todo-app:$TRAVIS_COMMIT-test .
+      # - docker build --target test_e2e --tag $DOCKER_USERNAME/todo-app:$TRAVIS_COMMIT-test_e2e .
+      - docker run -e SECRET_KEY -e USER_NAME -e PASSWORD -e DEFAULT_DATABASE -e LOGIN_DISABLED -e MONGO_CONNECTION_STRING -e APP_LOGIN_DISABLED -e GITHUB_CLIENT_ID -e GITHUB_SECRET -e USERS_ROLE --mount type=bind,source="$(pwd)"/todo_app,target=/project/todo_app $DOCKER_USERNAME/todo-app:$TRAVIS_COMMIT-test
+      # - docker run -e SECRET_KEY -e USER_NAME -e PASSWORD -e DEFAULT_DATABASE -e LOGIN_DISABLED -e MONGO_CONNECTION_STRING -e APP_LOGIN_DISABLED -e GITHUB_CLIENT_ID -e GITHUB_SECRET -e USERS_ROLE --mount type=bind,source="$(pwd)"/todo_app,target=/project/todo_app $DOCKER_USERNAME/todo-app:$TRAVIS_COMMIT-test_e2e
```

Comment:
Regarding your e2e test not working - you are trying to pass in `-e MONGO_CONNECTION_STRING`, but that environment variable isn't set in Travis.

You can see the list of environment variables at the top of the logs. 
https://app.travis-ci.com/github/polylearner/DevOps-Course-Starter/builds/235391566

Maybe you were trying to rely on the `MONGO_CONNECTION_STRING = ...` line in the `before_deploy` section, but that currently runs **after** the `script` section

Also if login is disabled, you shouldn't need a real github client ID/secret.